### PR TITLE
Initial ticket compression with zstd

### PR DESCRIPTION
**Pull request by pjenys0**
Add ticket compression with zstd.
TBD:
- not yet test
- met testing issue with current test : `fatal error: in "rpc_wallet_tests/rpc_z_shieldcoinbase_parameters": std::runtime_error: could not load param file at /home/nvnguyen/.pastel-params/sprout-verifying.key`

### CODE REVIEW

**Comment by akobrin1** [comment]
Could you please fix the PR build for Windows & macOS?

**Comment by pjenys0** [comment]
> Could you please fix the PR build for Windows & macOS?

@akobrin1 : fixed it, could you have a look?

**Review by a-ok123 — DISMISSED** [review]

File: src/mnode/ticket-processor.h [hunk 1]
```diff
@@ -13,6 +13,8 @@
 #include "mnode/ticket.h"
 
 constexpr int DATASTREAM_VERSION = 1;
+constexpr unsigned char kTicketCompressEnableMask  = 0x80;
+constexpr unsigned char kTicketCompressDisableMask = 0x7f;
```
Comment (akobrin1) [inline src/mnode/ticket-processor.h:17]: can you please uppercase global constants and also make second one depend on the first one like this:
constexpr unsigned char TICKET_COMPRESS_ENABLE_MASK  = 0x80;
constexpr unsigned char TICKET_COMPRESS_DISABLE_MASK = ~TICKET_COMPRESS_ENABLE_MASK;

File: src/mnode/ticket-processor.cpp [hunk 2]
```diff
@@ -132,18 +133,30 @@ bool CPastelTicketProcessor::UpdateDB(CPastelTicket &ticket, string& txid, const
...
+    compressed = (u & kTicketCompressEnableMask) > 0;
+    ticket_id = static_cast<TicketID>(u & kTicketCompressDisableMask);
     return true;
 }
 
+void parsingTicketData(CDataStream& data_stream, CPastelTicket &ticket, const bool compressed) {
+    if (!compressed) {
+        data_stream >> ticket;
+    } else {
+        CDataStream decrompress_stream(SER_NETWORK, DATASTREAM_VERSION);
```
Comment (akobrin1) [inline src/mnode/ticket-processor.cpp:153]: decrompress... -> decompress_stream

File: src/mnode/ticket-processor.cpp [hunk 3]
```diff
@@ -152,10 +165,11 @@ bool CPastelTicketProcessor::ValidateIfTicketTransaction(const int nHeight, cons
     CMutableTransaction mtx(tx);
 
     std::string error_ret;
+    bool compressed;
```
Comment (akobrin1) [inline src/mnode/ticket-processor.cpp:168]: initialize to false

File: src/mnode/ticket-processor.cpp [hunk 4]
```diff
@@ -276,10 +290,11 @@ bool CPastelTicketProcessor::ValidateIfTicketTransaction(const int nHeight, cons
 bool CPastelTicketProcessor::ParseTicketAndUpdateDB(CMutableTransaction& tx, const unsigned int nBlockHeight)
 {
     std::string error_ret;
+    bool compressed;
```
Comment (akobrin1) [inline src/mnode/ticket-processor.cpp:311]: initialize to false

File: src/mnode/ticket-processor.cpp [hunk 5]
```diff
@@ -331,9 +346,10 @@ std::unique_ptr<CPastelTicket> CPastelTicketProcessor::GetTicket(const uint256 &
 
     TicketID ticket_id;
     std::string error_ret;
+    bool compressed;
```
Comment (akobrin1) [inline src/mnode/ticket-processor.cpp:367]: initialize to false

**Review by akobrin1 — CHANGES_REQUESTED** [review]

File: src/mnode/ticket-processor.cpp (hunk 3 above)
Comment (pjenys0) [reply src/mnode/ticket-processor.cpp:168]: thanks. this is my bad. a little bit still kept golang stuff in my head :).

File: src/mnode/ticket-processor.cpp [hunk 6]
```diff
@@ -783,9 +799,13 @@ std::string CPastelTicketProcessor::SendTicket(const CPastelTicket& ticket)
...
     const CAmount extraAmount = ticket.GetExtraOutputs(extraOutputs);
 
+    CDataStream ticket_stream(SER_NETWORK, DATASTREAM_VERSION);
+    ticket_stream << ticket;
+    CDataCompressor compressor(ticket_stream);
+
     CDataStream data_stream(SER_NETWORK, DATASTREAM_VERSION);
-    data_stream << (uint8_t)ticket.ID();
-    data_stream << ticket;
+    data_stream << (uint8_t)((uint8_t)ticket.ID() | TICKET_COMPRESS_ENABLE_MASK);
```
Comment (vadymeng0) [inline src/mnode/ticket-processor.cpp:807]: should use C++ static_cast instead of C-style cast

File: src/mnode/ticket-processor.cpp [hunk 7]
```diff
@@ -132,18 +133,30 @@ bool CPastelTicketProcessor::UpdateDB(CPastelTicket &ticket, string& txid, const
     return true;
 }
 
-bool preParseTicket(const CMutableTransaction& tx, CDataStream& data_stream, TicketID& ticket_id, std::string& error, bool log = true)
+bool preParseTicket(const CMutableTransaction& tx, CDataStream& data_stream, TicketID& ticket_id, std::string& error, bool& compressed, bool log = true)
 {
     vector<unsigned char> output_data;
     if (!CPastelTicketProcessor::ParseP2FMSTransaction(tx, output_data, error))
```
Comment (vadymeng0) [inline src/mnode/ticket-processor.cpp:139]: Do we need assign a value to `compressed` before returning?

**Review by VitaliyVK777 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**